### PR TITLE
Fix #1796: Nativelib toCString() & fromCString() now return nulls.

### LINKED ISSUE #1796: fromCString(null) returns java.lang.RuntimeException not NullPointerException
@david-bouyssie reported a real life problem he has been encountering.  We have been discussing 
it offline and refining possible solutions.  Time has come, I think, to open up the discussion.

I barfed as I was writing the `fromCString(null)` as I was writing the title and any people
probably averted their eyes and refused to read it.  The classical, and my usual answer,
is "Don't do that, If nulls are a possibility, say from a getenv() call, check your results and/or
errno".  Still a good, albeit unfriendly, response. The classical reply is"Nulls exist in the real world,
get over it. Library code should not be actively hostile in handling them."

To the current problem.  My recent understanding is that Scala practice is to not do explicit
checks for null and rely upon Scala throwing a NullPointerException, which can be caught,
Fine so far.  

Now scalanative.unsafe.fromCString is, well, unsafe. It is on the edge of Scala and OS code.
David reported that fromCString(null) returns java.lang.RuntimeException.  He is innocent
of bad coding; he is working with an API which deals with nulls as default values.
Investigation shows that the generated binary for null.getBytes(mumble) throws the
Unix signal 11 Segmentation fault. Scala adds 128 and reports it as java.lang.RuntimeException 139,
which can not be caught.  Not friendly for a recoverable error (yes, people should check rather
than relying upon try/catch).  The expected/desired behavior would be to get a NullPointerException.

There are a couple of possible solutions.

1) Do nothing, we have other things to do. 

2) Document as a "Do not do that", probably in the clib(?) documentation, where fromCString()
    is described.  Suggest using a wrapper around the fromCString() call and possibly
    provide an example.

3) add the lines of code roughly so that the usual, expected, & NullPointerException, which can be
    caught.
```
if (cstr == null) {
    throw new NullPointerException
}
```
   This adds a test and branch across a small code block for each and every call to the
    method, but is friendlier and more inline with norms & practice.

4) Implement a fromCStringOrElse(str, default = TBS) method. This would be the example
    suggested above, but better.  Similar to Option.getOrElse, it would return the default
    in case of null arg or other.  This is adds an extra entrypoint to the SN unsafe API
    doing work which developers should be doing themselves but it avoids the support
    cost of a number of devos writing their own, possibly wrong, code.  That is, it is
    friendlier, but not without cost.

    The default for the default is open to discussion, if the discussion gets that far. A possibility
     would be the empty string.  Another possibility is null in, null out.  

Thoughts? Beyond the very appropriate yeech!

### PR DESCRIPTION
**Pull request by LeeTibbert**
>    In the sweat of thy face shalt thou eat bread, till thou return unto
>   the ground; for out of it wast thou taken: for null thou art, and
>   unto null shalt thou return. [1]

  * This PR was motivated by Issue #1796.  There is extensive discussion
    of design alternatives in that Issue. This PR implements the
    final decision; toCString() & fromCString are to
    return null when given a null.

  * Two test cases for exercising null handling were added to
    were added to CStringSuite.scala.

    To isolate the changes of this PR, I left CStringSuite.scala
    as the old-style Scala Native test Suite it was before I touched
    it this time.  If/when the core changes of this PR are found
    acceptable, I can sweep through and convert that file to JUnit,
    under the guidance of: You touch it, you own it.

Documentation:

  * The standard changelog entry is requested.

  * I added a one sentence paragraph to `docs/user/interop.rst`
    to make the newly defined behavior accessible to users of Scala Native.

    `interop.rst` had a number (7 or 8) warnings before I touched it.
    I left it with the same warnings.  That is, I did not introduce
    any warning, but, to keep this PR reviewable on-sight, I did
    not fix any.  I think I have a PR in the queue which fixes some
    of those warnings and others; yielding, at that time, a clean build.

Testing:

  Safety -

  + Built and tested ("test-all") in debug mode using sbt 1.3.13 on
      X86_64 only . All tests pass.

  Efficacy -

  + The two newly cases in CStringSuite.scala to exercise the changes
      of this both passed.

  Documentation -

  + I did a successful manual html build of the `doc` directory.

[1]:  Adapted from [original](https://www.phrases.org.uk/meanings/ashes-to-ashes.html)

### CODE REVIEW

File: nativelib/src/main/scala/scala/scalanative/unsafe/package.scala [hunk 1]
```diff
@@ -140,16 +140,20 @@ package object unsafe {
...
                   charset: Charset = Charset.defaultCharset()): String = {
-    val len   = libc.strlen(cstr).toInt
-    val bytes = new Array[Byte](len)
-
-    var c = 0
-    while (c < len) {
-      bytes(c) = !(cstr + c)
-      c += 1
+    if (cstr == null.asInstanceOf[CString]) {
+      null.asInstanceOf[String]
```
Comment (sjrd) [inline nativelib/src/main/scala/scala/scalanative/unsafe/package.scala:144]: I'm pretty sure the `asInstanceOf`s are not necessary:
```suggestion
    if (cstr == null) {
      null
```

File: nativelib/src/main/scala/scala/scalanative/unsafe/package.scala [hunk 2]
```diff
@@ -161,18 +165,22 @@ package object unsafe {
   /** Convert a java.lang.String to a CString using given charset and allocator.
    */
   def toCString(str: String, charset: Charset)(implicit z: Zone): CString = {
-    val bytes = str.getBytes(charset)
-    val cstr  = z.alloc(bytes.length + 1)
+    if (str == null.asInstanceOf[String]) {
+      null.asInstanceOf[CString]
```
Comment (sjrd) [inline nativelib/src/main/scala/scala/scalanative/unsafe/package.scala:169]: Same here:
```suggestion
    if (str == null) {
      null
```

File: unit-tests/src/test/scala/scala/scalanative/unsafe/CStringSuite.scala [hunk 3]
```diff
@@ -62,6 +62,11 @@ object CStringSuite extends tests.Suite {
     assertEquals("\\'", fromCString(c"\\'"))
   }
 
+  // Issue 1796
+  test("fromCString(null) returns null") {
+    assertNull(fromCString(null.asInstanceOf[CString]))
```
Comment (sjrd) [inline unit-tests/src/test/scala/scala/scalanative/unsafe/CStringSuite.scala:67]:
```suggestion
    assertNull(fromCString(null))
```

File: unit-tests/src/test/scala/scala/scalanative/unsafe/CStringSuite.scala [hunk 4]
```diff
@@ -73,6 +78,11 @@ object CStringSuite extends tests.Suite {
     assert(szTo.charAt(3) == '4')
   }
 
+  // Issue 1796
+  test("toCString(null) return null") {
+    Zone { implicit z => assertNull(toCString(null.asInstanceOf[String])) }
```
Comment (sjrd) [inline unit-tests/src/test/scala/scala/scalanative/unsafe/CStringSuite.scala:83]:
```suggestion
    Zone { implicit z => assertNull(toCString(null)) }
```

**Review by sjrd — CHANGES_REQUESTED** [review]
Thanks for the PR. There are several `asInstanceOf`s that are not necessary, I believe.

File: nativelib/src/main/scala/scala/scalanative/unsafe/package.scala (hunk 1 above)
Comment (LeeTibbert) [reply nativelib/src/main/scala/scala/scalanative/unsafe/package.scala:144]: Thank you for the timely review & the suggested fixes.  Does asInstanceOf have a runtime cost?

**Review by sjrd — APPROVED** [review]

**Comment by LeeTibbert** [comment]
@david-bouyssie, @jokade   Does this PR provide what you were describing & needing to ease your work?
Did I capture the idea?

**Review by ekrich — APPROVED** [review]
LGTM now.

**Pull request merged**